plugins/alembic/module/vtkF3DAlembicReader.cxx

```diff
@@ -20,51 +22,338 @@
 #pragma warning(pop)
 #endif
 
+using IndicesContainer = std::vector<int>;
+using V3fContainer = std::vector<Alembic::Abc::V3f>;
+using AttributesContainer = std::map<std::string, V3fContainer>;
+using PerFaceWavefrontIndicesTripletsContainer = std::vector<Alembic::Abc::V3i>;
+using PerMeshWavefrontIndicesTripletsContainer =
+  std::vector<PerFaceWavefrontIndicesTripletsContainer>;
+
+constexpr int pIndicesOffset = 0;
+constexpr int uvIndicesOffset = 1;
+constexpr int nIndicesOffset = 2;
+
+struct IntermediateGeometry
+{
+  AttributesContainer Attributes;
+  PerMeshWavefrontIndicesTripletsContainer Indices;
+  bool uvFaceVarying = false;
+  bool nIsFaceVarying = false;
+};
+
 class vtkF3DAlembicReader::vtkInternals
 {
+  void SetupIndicesStorage(const Alembic::AbcGeom::Int32ArraySamplePtr& faceVertexCounts,
+    PerMeshWavefrontIndicesTripletsContainer& extractedIndices)
+  {
+
+    for (size_t i = 0; i < faceVertexCounts->size(); i++)
+    {
+      extractedIndices.emplace_back(faceVertexCounts->get()[i]);
+    }
+  }
+
+  template<typename I>
+  void UpdateIndices(const I& attributeIndices, int indicesOffset,
+    PerMeshWavefrontIndicesTripletsContainer& meshIndices, bool doReverseRotate = true)
+  {
+    size_t faceIndicesCounter = 0;
+    for (auto& perFaceIndices : meshIndices)
+    {
+      // Perform the collection first
+      size_t thisFaceVertexCount = perFaceIndices.size();
+      IndicesContainer thisFaceIndices;
+      for (size_t j = 0; j < thisFaceVertexCount; j++)
+      {
+        auto vertex = attributeIndices->get()[faceIndicesCounter];
+        thisFaceIndices.emplace_back(vertex);
+        faceIndicesCounter++;
+      }
+      if (doReverseRotate)
+      {
+        std::reverse(thisFaceIndices.begin(), thisFaceIndices.end());
+        std::rotate(thisFaceIndices.begin(), thisFaceIndices.begin() + thisFaceIndices.size() - 1,
+          thisFaceIndices.end());
+      }
+
+      // Now update the mesh's indices
+      for (size_t j = 0; j < thisFaceVertexCount; j++)
+      {
+        perFaceIndices[j][indicesOffset] = thisFaceIndices[j];
+      }
+    }
+  }
+
+  void PointDuplicateAccumulator(
+    const IntermediateGeometry& originalData, IntermediateGeometry& duplicatedData)
+  {
+    bool needToDuplicate = originalData.uvFaceVarying || originalData.nIsFaceVarying;
+
+    auto uvMapIter = originalData.Attributes.find("uv");
+    auto nMapIter = originalData.Attributes.find("N");
+    bool haveUV = uvMapIter != originalData.Attributes.end();
+    bool haveN = nMapIter != originalData.Attributes.end();
+
+    if (needToDuplicate)
+    {
+      auto faceCount = originalData.Indices.size();
+      duplicatedData.Indices.resize(faceCount);
+      for (size_t i = 0; i < faceCount; i++)
+      {
+        auto thisFaceVertexCount = originalData.Indices[i].size();
+
+        // TODO check with nYue
+        duplicatedData.Indices[i].resize(
+          thisFaceVertexCount, Alembic::Abc::V3i(-9999, -9999, -9999));
```

Comment:
@nyue what does -9999 means ?

Reply:
I was originally thinking that initialising it with some known sentinel value makes it easier to debug and when I see -9999, I know I have miscalculated what needs to be initialise with proper value down the line. Feel free to change it if this defensive programming approach does not have sufficient merit.